src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1401,20 +1661,97 @@ class StickyClassDumper : public KlassClosure {
   }
 };
 
+// Large object heap dump support.
+// To avoid memory consumption, when dumping large objects such as huge array and
+// large objects whose size are larger than LARGE_OBJECT_DUMP_THRESHOLD, the scanned
+// partial object/array data will be sent to the backend directly instead of caching
+// the whole object/array in the internal buffer.
+// The HeapDumpLargeObjectList is used to save the large object when dumper scans
+// the heap. The large objects could be added (push) parallelly by multiple dumpers,
+// But they will be removed (popped) serially only by the VM thread.
+class HeapDumpLargeObjectList : public CHeapObj<mtInternal> {
+ private:
+  class HeapDumpLargeObjectListElem : public CHeapObj<mtInternal> {
+   public:
+    HeapDumpLargeObjectListElem(oop obj) : _obj(obj), _next(NULL) { }
+    oop _obj;
+    HeapDumpLargeObjectListElem* _next;
+  };
+
+  HeapDumpLargeObjectListElem* _head;
+  uint _length;
+
+ public:
+  HeapDumpLargeObjectList() : _head(NULL),  _length(0) { }
+
+  void atomic_push(oop obj) {
+    assert (obj != NULL, "sanity check");
+    HeapDumpLargeObjectListElem* entry = new HeapDumpLargeObjectListElem(obj);
+    if (entry == NULL) {
+      warning("failed to allocate element for large object list");
+      return;
+    }
+    assert (entry->_obj != NULL, "sanity check");
+    while (true) {
+      HeapDumpLargeObjectListElem* old_head = Atomic::load_acquire(&_head);
+      HeapDumpLargeObjectListElem* new_head = entry;
+      if (Atomic::cmpxchg(&_head, old_head, new_head) == old_head) {
+        // successfully push
+        new_head->_next = old_head;
+        Atomic::inc(&_length);
+        assert(new_head->_obj == obj, "must equal");
+        return;
+      }
+    }
+  }
+
+  oop pop() {
+    if (_head == NULL) {
+      assert(_length == 0, "sanity check");
+      return NULL;
+    }
+    HeapDumpLargeObjectListElem* entry = _head;
+    if (_head->_next != NULL) {
+      _head = _head->_next;
+    }
+    entry->_next = NULL;
```

Comment:
I don't see why you would NULL out the _next field. The entry is deleted shortly after anyway.